receiver/awsfirehosereceiver/unmarshaler/cwmetricstream/metricsbuilder.go

```diff
@@ -0,0 +1,157 @@
+// Copyright  The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cwmetricstream // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsfirehosereceiver/unmarshaler/cwmetricstream"
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"go.opentelemetry.io/collector/model/pdata"
+	conventions "go.opentelemetry.io/collector/model/semconv/v1.5.0"
+)
+
+const (
+	attributeAWSCloudWatchMetricStreamName = "aws.cloudwatch.metric_stream_name"
+	dimensionInstanceID                    = "InstanceId"
+	namespaceDelimiter                     = "/"
+)
+
+type resourceMetricsBuilder struct {
+	metricStreamName string
+	accountID        string
+	region           string
+	namespace        string
+	metricBuilders   map[string]*metricBuilder
+}
+
+func newResourceMetricsBuilder(metricStreamName, accountID, region, namespace string) *resourceMetricsBuilder {
+	return &resourceMetricsBuilder{
+		metricStreamName: metricStreamName,
+		accountID:        accountID,
+		region:           region,
+		namespace:        namespace,
+		metricBuilders:   make(map[string]*metricBuilder),
+	}
+}
+
+// AddMetric adds a metric to one of the metric builders based on
+// the key generated for each.
+func (rmb *resourceMetricsBuilder) AddMetric(metric cWMetric) {
+	metricKey := rmb.toMetricKey(metric)
+	mb, ok := rmb.metricBuilders[metricKey]
+	if !ok {
+		mb = newMetricBuilder(metric.MetricName, metric.Unit)
+		rmb.metricBuilders[metricKey] = mb
+	}
+	mb.AddDataPoint(metric)
+}
+
+// Build creates the pdata.ResourceMetrics from the added metrics.
+func (rmb *resourceMetricsBuilder) Build() pdata.ResourceMetrics {
+	rm := pdata.NewResourceMetrics()
+	ilm := rm.InstrumentationLibraryMetrics().AppendEmpty()
+	rmb.toResource().CopyTo(rm.Resource())
+	for _, mb := range rmb.metricBuilders {
+		mb.Build().CopyTo(ilm.Metrics().AppendEmpty())
```

Comment:
```suggestion
	rmb.toResource(rm.Resource())
	for _, mb := range rmb.metricBuilders {
		mb.Build(ilm.Metrics().AppendEmpty())
```

By taking the target resource/metric directly we should be able to avoid copying.